### ISSUE TITLE
Getting "The number of sequence start flags does not match the number of sequences" Error

### ISSUE THREAD

**Issue by arduano** [Microsoft/CNTK#1961]
I was attempting to make a GAN LSTM, mostly copying my previous working code for a char LSTM.
I am getting this error:
`Value::Create:: The number (1) of sequence start flags does not match the number (200) of sequences,`
at
`D_trainer.train_minibatch(arguments)`

My train method:
```Python
def make_model_and_train(data_root_path, model_root_path, epochs, max_num_minibatches, name):
    data, char_to_ix, ix_to_char, vocab_dim, data_dim = load_data_and_vocab(os.path.join(data_root_path, name))

    batch_axis = ct.Axis.default_batch_axis()
    input_seq_axis = ct.Axis('inputAxis')

    model = create_model(vocab_dim, lstm_dims)
    model_D = create_model(vocab_dim, lstm_dims_D, True)

    input_dynamic_axes = [batch_axis, input_seq_axis]
    Z = ct.ops.sequence.input_variable(noise_shape, sequence_axis=input_seq_axis)
    G_real = ct.ops.sequence.input_variable(vocab_dim, sequence_axis=input_seq_axis)
    G_fake = model(Z)
    D_fake = model_D(G_fake)
    D_real = D_fake.clone(
        method = 'share',
        substitutions = {G_fake.output: G_real}
    )


    G_loss = 1.0 - ct.log(D_fake)
    D_loss = -(ct.log(D_real) + ct.log(1.0 - D_fake))

    minibatches_per_epoch = min(data_dim // minibatch_dim, max_num_minibatches // epochs)
    
    lt_per_sample = ct.learning_rate_schedule([(30, 0.001),(100, 0.0005)], ct.UnitType.sample, minibatches_per_epoch)
    lt_per_sample_d = ct.learning_rate_schedule([(30, 0.001),(100, 0.0005)], ct.UnitType.sample, minibatches_per_epoch)

    clipping_threshold_per_sample = 3
    gradient_clipping_with_truncation = True

    G_learner = ct.momentum_sgd(G_fake.parameters ,lt_per_sample, ct.momentum_as_time_constant_schedule(1100),gradient_clipping_threshold_per_sample = clipping_threshold_per_sample, gradient_clipping_with_truncation = gradient_clipping_with_truncation)
    D_learner = ct.momentum_sgd(D_real.parameters ,lt_per_sample_d, ct.momentum_as_time_constant_schedule(1100),gradient_clipping_threshold_per_sample = clipping_threshold_per_sample, gradient_clipping_with_truncation = gradient_clipping_with_truncation)

    pp_G = ct.logging.ProgressPrinter(100)
    pp_D = ct.logging.ProgressPrinter(100)

    G_trainer = ct.Trainer(
        G_fake,
        (G_loss, None),
        G_learner,
        pp_G
    )
    D_trainer = ct.Trainer(
        D_real,
        (D_loss, None),
        D_learner,
        pp_D
    )
    
    sample_freq = 4000

    print ("Running %d epochs with %d minibatches per epoch" % (epochs, minibatches_per_epoch))
    print('')

    s = sample(G_fake, ix_to_char)
    print(s)

    for e in range(epochs):
        mask = [True]
        for b in range(minibatches_per_epoch):
            real = get_data(b, minibatch_dim, data, char_to_ix, vocab_dim)
            Z_data = noise_sample(minibatch_dim)
            arguments = ({G_real: real, Z: Z_data}, mask)
            D_trainer.train_minibatch(arguments)

            arguments = ({Z: Z_data}, mask)
            G_trainer.train_minibatch(arguments)

            pp_G.update_with_trainer(G_trainer)
            pp_D.update_with_trainer(D_trainer)


            mask = [False]
            
            global_minibatch = e * minibatches_per_epoch + b
            if global_minibatch % sample_freq == 0:
                s = sample(G_fake, ix_to_char)
                print(s)
            if global_minibatch % (sample_freq * 10) == 0:
                s = sample(G_fake, ix_to_char, 5000)
                print(s)
                f = open('samples/sample_from_%g.txt'%(e*minibatches_per_epoch + b), 'w', encoding='utf-8')
                f.write(s)
                f.close()
        model_filename = '%s/%s/%s_epoch_%g.dnn' % (model_root_path, name, name, e+1)
        G_trainer.save_model(model_filename)
        print("Saved model to '%s'" % model_filename)
        model_filename = '%s/%s/%s_epoch_%d.dnn' % (model_root_path, name, name, e+1)
        D_trainer.save_model(model_filename)
        print("Saved model to '%s'" % model_filename)
```

**Comment by sayanpa** [comment]
I was wondering if you have tried with the network with keeping the input_variable to accept the sequence axis to be the default dynamic axes. Currently you are passing input_seq_axis = ct.Axis('inputAxis').

**Comment by arduano** [comment]
I'm not exactly sure what you mean, but I have tried many options from other LSTMs that I found. Oh, and I could not find any direct solution online.

I also tried to make another basic LSTM later on, and had the same problem, even though the code was mostly copied (much closer to the original). I ended up re-programming it in Keras.

**Comment by sayanpa** [comment]
Have you tried your network with CNTK set as the backend of Keras. https://docs.microsoft.com/en-us/cognitive-toolkit/using-cntk-with-keras

Once you get it working with Keras. We can translate the model to CNTK front end. With LSTM you should see a speed up with CNTK set as backend. Moving to CNTK API should be very straight forward since our API's are close to Keras. 

Let us know how the Keras API with CNTK as backend goes. Will keep this issue open for now since we continue to work through it.

**Comment by arduano** [comment]
It didn't go as well as I wanted because the CNTK backend doesn't support stateful LSTMs. But I did end up finding a way around that and training it without using stateful, and then sampling with stateful using the TF backend.

Why doesn't Keras support stateful with CNTK? The performance boost is impressive, but most of my LSTM projects work best with stateful models.

**Comment by sayanpa** [comment]
Can you share any one of your simpler stateful models in Keras that you are referring to? May be as a download link or via email if you prefer.

**Comment by arduano** [comment]
[Keras_char_LSTM.zip](https://github.com/Microsoft/CNTK/files/1201992/Keras_char_LSTM.zip)
I tried re-creating my char LSTM in Keras

**Comment by souptc** [comment]
@arduano , with CNTK 2.1 release, we do support stateful lstm in keras. Please update your keras to latest in keras master branch, and install CNTK v2.1.

I tried your script, it seems works with latest Keras and CNTK:
`
9998 / 10000
9999 / 10000
Itt:
! now the Prince the !`

except one thing: in model "sample.json", for the LSTM layer, you should set the unroll to false:
`return_state": false, "go_backwards": false, "stateful": true, "unroll": true`
Because the time step axis is dynamic, unroll is not supported on dynamic axis, both for cntk or other frameworks like tensorflow.

**Comment by arduano** [comment]
Great. So as long as I use Keras, I'm fine. Thanks!
